data_pipeline/generic_web_api/generic_web_api_data_etl.py

```diff
@@ -0,0 +1,448 @@
+from data_pipeline.generic_web_api.generic_web_api_config import (
+    WebApiConfig
+)
+
+
+# pylint: disable=too-few-public-methods
+class ModuleConstant:
+    DEFAULT_TIMESTAMP_FORMAT = "%Y-%m-%d %H:%M:%S%z"
+    BQ_SCHEMA_FIELD_NAME_KEY = "name"
+    BQ_SCHEMA_SUBFIELD_KEY = "fields"
+    BQ_SCHEMA_FIELD_TYPE_KEY = "type"
+
+
+def get_timestamp_as_string(
+        timestamp: datetime,
+        timestamp_format: str = ModuleConstant.DEFAULT_TIMESTAMP_FORMAT):
+    return timestamp.strftime(
+        timestamp_format
+    )
+
+
+def get_current_timestamp_as_string():
+    return datetime.now(
+        timezone.utc
+    ).strftime(ModuleConstant.DEFAULT_TIMESTAMP_FORMAT)
+
+
+def parse_timestamp_from_str(timestamp_as_str, time_format: str = None):
+    if time_format:
+        timestamp_obj = datetime.strptime(
+            timestamp_as_str.strip(), time_format
+        )
+    else:
+        timestamp_obj = dateparser.parse(
+            timestamp_as_str.strip()
+        )
+    return timestamp_obj
+
+
+def get_stored_state(
+        data_config: WebApiConfig,
+):
+    try:
+        stored_state = (
+            parse_timestamp_from_str(
+                download_s3_object_as_string(
+                    data_config.state_file_bucket_name,
+                    data_config.state_file_object_name
+                ),
+                ModuleConstant.DEFAULT_TIMESTAMP_FORMAT
+            )
+            if data_config.state_file_bucket_name and
+            data_config.state_file_object_name else None
+        )
+
+    except ClientError as ex:
+        if ex.response['Error']['Code'] == 'NoSuchKey':
+            stored_state = parse_timestamp_from_str(
+                data_config.default_start_date,
+                data_config.url_manager.date_format
+            ) if (
+                data_config.default_start_date and
+                data_config.url_manager.date_format
+            ) else None
+
+        else:
+            raise ex
+    return stored_state
+
+
+def get_newline_delimited_json_string_as_json_list(json_string):
+    return [
+        json.loads(line) for line in json_string.split("\n")
+        if line.strip()
+    ]
+
+
+# pylint: disable=fixme,too-many-arguments
+def get_data_single_page(
+        data_config: WebApiConfig,
+        cursor: str = None,
+        from_date: datetime = None,
+        until_date: datetime = None,
+        page_number: int = None,
+) -> (str, dict):
+    url = data_config.url_manager.get_url(
+        from_date=from_date,
+        to_date=until_date,
+        cursor=cursor,
+        page_number=page_number,
+    )
+
+    with requests_retry_session() as session:
+        if (
+                data_config.authentication and
+                data_config.authentication.authentication_type == "basic"
+        ):
+            session.auth = tuple(data_config.authentication.auth_val_list)
+        session.verify = False
+        session_request = session.get(url)
```

Comment:
I would maybe rename the variables. I believe this one should be a `response` (by the time the `get` method returned we will already have the response, unless the response was streamed). Whereas `resp` could maybe make clear that it's the content. e.g. `response_content`.